doc/sphinx/language/gallina-extensions.rst

```diff
@@ -1105,12 +1105,17 @@ wrapped into the :g:`Float64.of_float` constructor, e.g.:
 Bidirectionality hints
 ----------------------
 
-When type-checking an application, Coq normally does not use information from
-the context to infer the types of the arguments. It only checks after the fact
-that the type inferred for the application is coherent with the expected type.
-Bidirectionality hints make it possible to specify that after type-checking the
-first arguments of an application, typing information should be propagated from
-the context to help inferring the types of the remaining arguments.
+Two common methods for type-checking are:
+
+* verifying that a construct matches a known type, and
+* inferring the type of a construct by analyzing the construct.
+
+Methods that combine these approaches are known as *bidirectional typing*.
+Coq normally uses only the first approach to infer the types of arguments,
+then later verifies that the inferred type is consistent with the expected type.
+*Bidirectionality hints* specify to use both methods: after type-checking the
+first arguments of an application (as defined with `&` in :cmd:`Arguments`),
+typing information from them is propagated to the remaining arguments to help infer their types.
```

Comment:
Usually, we call type-checking the first bullet point and type inference the second one. In many cases, Coq does type inference first and checks the type afterwards, and gives an error if it doesn't match. It is clear that in the case of bidirectionality hints, Coq will use the information that is available that it wouldn't have used otherwise. However, I am not sure of how accurate your description is and I would love if it could be checked by an expert.

In particular, I am puzzled by "Coq normally uses only the first approach to infer the types of arguments". It looks like Coq is using type-checking to do type-inference by default. Did you meant to write "the second approach"?

Reply:
I based what I wrote on my reading of the old language here.  The two methods sound distinct, but I would have trouble giving good concrete examples distinguishing how Coq uses the two.  I figured an explicit declaration of the type of a parameter would allow type checking while inferring a type based on a function call would be inference (e.g. in OCaml, calling List.mem implies the argument is some kind of list).  Hand-waving.

Perhaps you can get an expert to comment?

Reply:
Let's ask @JasonGross.

Reply:
The problem with this description is that it refers to a single construct, while bidi is inherently about using the typechecking information of some parts to do inference on other parts.  A typical example, IIRC, is that if you have `Record foo (b : bool) := {| bar : if b then nat else bool |}` then `{| bar := true |} : foo false` will fail to typecheck because Coq says "no" when asked if `true` has type `if ?e then nat else bool`.  Enabling bidi causes Coq to first infer `?e` from `foo false` before doing the inner checking.  It think it really doesn't make sense to talk about bidi without talking about direction and giving more details on Coq's algorithm.  For example, I think that which problems need bidi depends on whether the underlying algorithm is top-down or bottom-up, though I'm not entirely sure.

Reply:
OK thanks @JasonGross. @jfehrle let's not spend too much time on this and let's request someone who can to provide a detailed description of the typing algorithm.